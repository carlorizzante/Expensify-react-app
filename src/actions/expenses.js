import uuid from 'uuid';
import database from '../firebase/firebase.js';

// ADD_EXPENSE
export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
  }
}

// Connect Firebase, store data, dispatch "addExpense"
export const startAddExpense = (expense) => {
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expense;

  return (dispatch) => {
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses')
      .push(expense)
      .then((ref) => dispatch(addExpense({
        id: ref.key,
        ...expense
      })));
  }
}

// REMOVE_EXPENSE
export const removeExpense = id => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
}
// EDIT_EXPENSE
export const editExpense = (id, updates = {}) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updates
  }
}
