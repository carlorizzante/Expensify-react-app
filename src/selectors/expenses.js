import moment from 'moment';

// January 1st 1970 -> Unix Epoch
// timestamp in milliseconds

// Get Visible Espenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter(expense => {
    // const createdAt = expense.createdAt;
    const createdAtMoment = moment(expense.createdAt);
    // const startDateMatch = typeof startDate !== 'number' || createdAt >= startDate;
    // const endDateMatch = typeof endDate !== 'number' || createdAt <= endDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  }).sort((exp1, exp2) => {
    if (sortBy == 'amount') return exp1.amount < exp2.amount;
    else return exp1.createdAt < exp2.createdAt;
  });
}
