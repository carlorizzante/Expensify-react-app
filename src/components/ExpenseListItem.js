import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';
import numeral from 'numeral';
import { startRemoveExpense } from '../actions/expenses.js';

export const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => {
  // const handleRemove = _ => dispatch(startRemoveExpense(id));
  return (
    <Link
      className="list-item"
      to={ `/edit/${id}` }
    >
      <div>
        <h3>{ description }</h3>
        <i>{ moment(createdAt).format("Do MMMM YYYY") }</i>
      </div>
      <h3>{ numeral(amount / 100).format('$0,00.00') }</h3>
      {/* <button onClick={ handleRemove }>Remove</button> */}
    </Link>
  );
}

export default connect(null)(ExpenseListItem);
