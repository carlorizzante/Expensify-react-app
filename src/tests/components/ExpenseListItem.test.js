import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem.js';
import expenses from '../fixtures/expenses.js';

test('should render the expense', () => {
  const expense = {
    dispatch: () => {},
    ...expenses[1]
  }
  const wrapper = shallow(<ExpenseListItem { ...expense }/>);
  expect(wrapper).toMatchSnapshot();
});
