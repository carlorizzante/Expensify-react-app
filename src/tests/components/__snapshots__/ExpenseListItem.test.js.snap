// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/bc23"
>
  <div>
    <h3>
      A bit later
    </h3>
    <i>
      22nd December 1969
    </i>
  </div>
  <h3>
    $2.00
  </h3>
</Link>
`;
